Mathlib/Order/KrullDimension.lean

```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
 Coheight of an element `a` of a pre-ordered set `α` is the Krull dimension of the subset `[a, +∞)`
 -/
 noncomputable def coheight (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Ici a)
+
+end definitions
+
+namespace krullDim
+
+variable {α β : Type _}
+
+variable [Preorder α] [Preorder β]
+
+lemma nonneg_of_nonempty [Nonempty α] : 0 ≤ krullDim α :=
+  le_sSup ⟨⟨0, fun _ ↦ @Nonempty.some α inferInstance, fun f ↦ f.elim0⟩, rfl⟩
+
+lemma eq_bot_of_isEmpty [IsEmpty α] : krullDim α = ⊥ := WithBot.ciSup_empty _
+
+lemma eq_top_of_infiniteDimensionalType [InfiniteDimensionalOrder α] :
+    krullDim α = ⊤ :=
+le_antisymm le_top <| le_iSup_iff.mpr <| fun m hm ↦ match m, hm with
+| ⊥, hm => False.elim <| by
+  haveI : Inhabited α := ⟨LTSeries.withLength _ 0 0⟩
+  exact not_le_of_lt (WithBot.bot_lt_coe _ : ⊥ < (0 : WithBot (WithTop ℕ))) <| hm default
+| some ⊤, _ => le_refl _
+| some (some m), hm => by
+  refine (not_lt_of_le (hm (LTSeries.withLength _ (m + 1))) ?_).elim
+  erw [WithBot.coe_lt_coe, WithTop.coe_lt_coe]
+  simp
+
+lemma eq_len_of_finiteDimensionalType [FiniteDimensionalOrder α] :
+    krullDim α = (LTSeries.longestOf α).length :=
+le_antisymm
+  (iSup_le <| fun _ ↦ WithBot.coe_le_coe.mpr <| WithTop.coe_le_coe.mpr <|
+    RelSeries.length_le_length_longestOf _ _) <|
+  le_iSup (fun (i : LTSeries _) ↦ (i.length : WithBot (WithTop ℕ))) <| LTSeries.longestOf _
+
+lemma eq_zero_of_unique [Unique α] : krullDim α = 0 :=  by
+  rw [eq_len_of_finiteDimensionalType (α := α), Nat.cast_eq_zero]
+  refine (LTSeries.longestOf_len_unique (default : LTSeries α) fun q ↦ show _ ≤ 0 from ?_).symm
+  by_contra r
+  exact ne_of_lt (q.step ⟨0, not_le.mp r⟩) <| Subsingleton.elim _ _
+
+/-- If `f : α → β` is a strictly monotonic function and `α` is an infinite dimensional type then so
+  is `β`. -/
+lemma infiniteDimensional_of_strictMono
+    (f : α → β) (hf : StrictMono f) [InfiniteDimensionalOrder α] :
+    InfiniteDimensionalOrder β :=
+  ⟨fun n ↦ ⟨(LTSeries.withLength _ n).map f hf, LTSeries.length_withLength α n⟩⟩
+
+lemma le_of_strictMono (f : α → β) (hf : StrictMono f) : krullDim α ≤ krullDim β :=
+  iSup_le <| fun p ↦ le_sSup ⟨p.map f hf, rfl⟩
+
+lemma height_mono {a b : α} (h : a ≤ b) : height α a ≤ height α b :=
+  le_of_strictMono (fun x ↦ ⟨x, le_trans x.2 h⟩) <| fun _ _ h ↦ h
```

Comment:
This should be earlier

Reply:
I moved this a bit earlier, but since it requires `krullDim_le_of_strictMono`, I can’t move it too early